src/load_balancer/cuda/cuda_collision_detection.cu

```diff
@@ -0,0 +1,231 @@
+    }
+  }
+}
+
+#define BUFFER_SIZE 8
+#define ELEMENT_SIZE 32
+#define BUFFER_SIZE_BITS ELEMENT_SIZE * BUFFER_SIZE
+
+__global__ void bitvector_to_position_list( 
+           size_t  ncubes, 
+           size_t  nspheres, 
+           size_t  LD_bit,
+    const int32_t* collisions, 
+    const int32_t* counts, 
+    const  size_t*  shell_size,
+          int32_t* position_list, 
+           size_t* nbe_list
+) {
+  __shared__ int32_t collisions_buffer[warp_size][warp_size][BUFFER_SIZE];
+
+  // We are converting a large number of small bitvectors into position lists. For this reason, I am assigning a single thread to each bitvector
+  // This avoids having to do popcounts and warp wide reductions, but hurts the memory access pattern
+
+  // All threads in a warp must be active to do shared memory loads, so we seperate out the threadId.x
+  for (int i_base = threadIdx.y * blockDim.x + blockIdx.x * blockDim.x * blockDim.y; i_base < ncubes; i_base += blockDim.x * blockDim.y * gridDim.x) {
+    const int i = i_base + threadIdx.x;
+    int32_t* out = position_list;
+    if (i != 0 && i < ncubes) {
+      out += counts[i-1];
+    } 
+
+    int current = 0;
+    size_t nbe = 0;
+    size_t nsphere_blocks = (nspheres + BUFFER_SIZE_BITS - 1) / BUFFER_SIZE_BITS;
+    for (int j_block = 0; j_block < nsphere_blocks; j_block++) {
+      // Each thread has a buffer of length BUFFER_SIZE. All the threads in the warp work to 
+      // load this data in a coalesced way (at least as much as possible)
+      for (int buffer_loop = 0; buffer_loop < warp_size; buffer_loop += warp_size/BUFFER_SIZE) {
+        const int t_id_x        = threadIdx.x % BUFFER_SIZE;
+        const int buffer_thread = threadIdx.x / BUFFER_SIZE;
+        const int buffer_idx    = buffer_thread + buffer_loop;
+        if (j_block * BUFFER_SIZE_BITS + t_id_x * ELEMENT_SIZE < nspheres && i_base + buffer_idx < ncubes) {
+          collisions_buffer[threadIdx.y][buffer_idx][t_id_x] = collisions[(i_base + buffer_idx) * LD_bit + j_block * BUFFER_SIZE + t_id_x];
+        }
+      }
+
+      __syncwarp();
+      if (i < ncubes) {  // Once the data has been loaded, we exclude the threads not corresponding to a bitvector
+        // We have loaded in BUFFER_SIZE_BITS elements to be processed by each warp
+        for (int j_inner = 0; j_inner < BUFFER_SIZE_BITS && j_block * BUFFER_SIZE_BITS + j_inner < nspheres; j_inner++) {
+          const int j = BUFFER_SIZE_BITS * j_block + j_inner;
+          const int j_int = j_inner / ELEMENT_SIZE;
+          const int j_bit = j_inner % ELEMENT_SIZE;
+          if( collisions_buffer[threadIdx.y][threadIdx.x][j_int] & (1 << (j_bit)) ) {
+            out[current++] = j;
+            nbe += shell_size[j];
+          }
+        }
+      }
+      __syncwarp();
+    }
+    if (i < ncubes) {
+      nbe_list[i] = nbe;
+    }
+  }
+}
+
+size_t compute_scratch( size_t ncubes, int32_t* counts_device ) {
+    // Compute scan of count vector and allocate memory
```

Comment:
Whoops, that comment is not correct. This function just computes the amount of temporary storage required and does not do any allocation. 

The actually allocation happens in `replicated_load_balancer.cxx`.  I did switch to using the CUB API instead of thrust because it is easier to control when the allocation of the temporary buffer happens